translation(JS): web/javascript/guide/modules (#646)
webdoky/content: files/uk/web/javascript/guide/modules/index.md

- `square.js` — містить:

  - `name` — сталу, що містить рядок 'square'.
  - `draw()` — малює на заданому полотні квадрат, що має задані розмір, положення й колір. Повертає об'єкт, що містить розмір, положення й колір квадрата.
  - `reportArea()` — вписує площу квадрата в особливий звітний список, отримавши його довжину.
  - `reportPerimeter()` — вписує периметр квадрата в особливий звітний список, отримавши його довжину.

### Крок убік — .mjs проти .js

Протягом цієї статті ми використовували для модульних файлів розширення `.js`, однак в інших ресурсах можна зустріти використання розширення `.mjs`. Наприклад, [документація V8 радить використовувати саме .mjs (англ.)](https://v8.dev/features/modules#mjs). Причини для цього:

- Покращення очевидності, тобто таке розширення робить очевидним те, які файли є модулями, а які – звичайним JavaScript.
- Певність щодо того, що модульні файли розбираються як модулі такими середовищами виконання, як [Node.js (англ.)](https://nodejs.org/api/esm.html#esm_enabling), й інструментами складання, як то [Babel (англ.)](https://babeljs.io/docs/en/options#sourcetype).

> Проте ми вирішили надалі використовувати `.js`, принаймні поки що. Аби модулі коректно працювали в браузері, треба пересвідчитися, що сервер видає їх з заголовком `Content-Type`, що містить MIME тип JavaScript, як то `text/javascript`. Без цього буде помилка строгої перевірки типу MIME і слова "Сервер видав тип MIME, що не є JavaScript", і браузер такий JavaScript не запускатиме. Більшість серверів самі по собі встановлюють коректний тип для файлів `.js`, але не для файлів `.mjs`. Серед серверів, що уже видають файли `.mjs` коректно – [GitHub Pages](https://pages.github.com/) і [`http-server`](https://github.com/http-party/http-server#readme) для Node.js.

Reviewer: Наскільки я знаю, рушій видає нелокалізовані помилки.
```suggestion
Проте ми вирішили надалі використовувати `.js`, принаймні поки що. Аби модулі коректно працювали в браузері, треба пересвідчитися, що сервер видає їх з заголовком `Content-Type`, що містить MIME тип JavaScript, як то `text/javascript`. Без цього буде помилка строгої перевірки типу MIME і слова "The server responded with a non-JavaScript MIME type", і браузер такий JavaScript не запускатиме. Більшість серверів самі по собі встановлюють коректний тип для файлів `.js`, але не для файлів `.mjs`. Серед серверів, що уже видають файли `.mjs` коректно – [GitHub Pages](https://pages.github.com/) і [`http-server`](https://github.com/http-party/http-server#readme) для Node.js.
```

Перш за все треба додати до елемента [`<script>`](/uk/docs/Web/HTML/Element/script) `type="module"`, аби оголосити сценарій як модуль. Для імпорту сценарію `main.js` застосовуємо таке:

```html
<script type="module" src="main.js"></script>
```

Також сценарій модуля можна вбудувати безпосередньо в файл HTML, розташувавши код JavaScript всередині тіла елемента `<script>`:

```html
<script type="module">
  /* Тут код модуля JavaScript */
</script>
```

Сценарій, в котрий імпортуються можливості модуля, по суті діє як модуль верхнього рівня. Якщо це упустити, то Firefox, наприклад, дає помилку "SyntaxError: import declarations may only appear at top level of a module".

Інструкції `import` і `export` можна використовувати лише в модулях, але не у звичайних сценаріях.

## Інші відмінності між модулями й звичайними сценаріями

- Слід звернути увагу на локальне тестування: якщо спробувати завантажити файл HTML локально (тобто з URL `file://`), то трапляться помилки CORS, у зв'язку з вимогами безпеки модулів JavaScript. Тестування треба проводити за допомогою сервера.

> - Крім того, зверніть увагу, що поведінка частин сценарію, визначених всередині модулів, коли порівняти зі звичайними сценаріями, може відрізнятися. Це пов'язано з тим, що модулі автоматично застосовують {{JSxRef("Strict_mode", "строгий режим", "", 1)}}.

Reviewer: ```suggestion
- Крім того, зверніть увагу, що поведінка частин сценарію, визначених всередині модулів, коли порівняти зі звичайними сценаріями, може відрізнятися. Це пов'язано з тим, що модулі автоматично застосовують {{JSxRef("Strict_mode", "суворий режим", "", 1)}}.
```

Визначені в модулі змінні обмежені модулем, якщо явно не прикріплені до глобального об'єкта. З іншого боку, глобально описані змінні доступні всередині модуля. Наприклад, в наступному коді:

```html
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8" />
    <title></title>
    <link rel="stylesheet" href="" />
  </head>
  <body>
    <div id="main"></div>
    <script>
      // Інструкція var створює глобальну змінну
      var text = "Привіт";
    </script>
    <script type="module" src="./render.js"></script>
  </body>
</html>
```

```js
/* render.js */
document.getElementById("main").innerText = text;
```

Сторінка покаже `Привіт`, адже глобальні змінні `text` і `document` – доступні в модулі. (Крім того, зауважте в цьому модулі, що модуль не обов'язково потребує інструкцій імпорту чи експорту: єдина необхідна річ – щоб точка входу мала `type="module"`.)

## Усталений експорт проти іменованого експорту

Експортована вище функціональність складалася з **іменованого експорту**: до кожної сутності (чи це функція, `const` тощо) було звертання за її ім'ям при експорті, і те саме ім'я використовувалося для звертання до неї також при імпорті.

> Також є різновид експорту, що зветься **усталеним експортом** — він розроблений для спрощення використання єдиної функції, наданої модулем, а також допомагає модулям JavaScript взаємодіяти з наявними модульними системами CommonJS та AMD (як це файно описано в [Заглибленні до ES6: Модулі (англ.)](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/) від Джейсона Орендорффа; шукайте "Default exports").

Reviewer: ```suggestion
Також є різновид експорту, що зветься **усталеним експортом** — він розроблений для спрощення використання єдиної функції, наданої модулем, а також допомагає модулям JavaScript взаємодіяти з наявними модульними системами CommonJS та AMD (як це файно описано в [Поглиблено про ES6: Модулі (англ.)](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/) від Джейсона Орендорффа; шукайте "Default exports").
```

Поки що, схоже, модулі з малюванням фігур на полотні працювали нормально. Але що буде, якщо спробувати додати модуль, котрий малює іншу фігуру, як то круг чи трикутник? Напевне, для таких фігур також були б відповідні функції, як то `draw()`, `reportArea()` тощо; якщо спробувати імпортувати інші функції з таким самим іменем в той самий модульний файл верхнього рівня, то будуть конфлікти й помилки.

На щастя, існує низка способів це обійти. Вони розглядаються в наступних розділах.

## Імпорт та експорт з перейменуванням

Всередині фігурних дужок інструкцій `import` і `export` можна використовувати ключове слово `as` вкупі з новим іменем можливості, аби змінити її ім'я, що використовуватиметься в модулі верхнього рівня.

Тож, наприклад, обидва наступні зразки роблять одне й те ж, хоч і в трохи різний спосіб:

```js
// у module.js
export { function1 as newFunctionName, function2 as anotherNewFunctionName };

// у main.js
import { newFunctionName, anotherNewFunctionName } from "./modules/module.js";
```

```js
// у module.js
export { function1, function2 };

// у main.js
import {
  function1 as newFunctionName,
  function2 as anotherNewFunctionName,
} from "./modules/module.js";
```

> Погляньмо на робочий приклад. В нашій директорії [renaming](https://github.com/mdn/js-examples/tree/master/module-examples/renaming) – така сама модульна система, як в попередньому прикладі, крім того, що додалися модулі `circle.js` і `triangle.js`, для малювання й звітування кругів та трикутників.

Reviewer: ```suggestion
Погляньмо на робочий приклад. В нашій директорії [renaming](https://github.com/mdn/js-examples/tree/master/module-examples/renaming) – така сама модульна система, як в попередньому прикладі, крім того, що додалися модулі `circle.js` і `triangle.js`, для малювання й звітування про круги та трикутники.
```

По-перше, палітра кольорів оголошується в окремому файлі [`colors.json`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/data/colors.json):

```json
{
  "yellow": "#F4D03F",
  "green": "#52BE80",
  "blue": "#5499C7",
  "red": "#CD6155",
  "orange": "#F39C12"
}
```

Потім створюється модуль на ім'я [`getColors.js`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/modules/getColors.js), котрий застосовує fetch-запит для завантаження файлу [`colors.json`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/data/colors.json) і повернення даних як об'єкта.

```js
// fetch-запит
const colors = fetch("../data/colors.json").then((response) => response.json());

export default await colors;
```

Зверніть увагу на останній рядок з експортом.

Ключове слово `await` застосовується до вказівки експортувати сталу `colors`. А отже – інші модулі, котрі включатимуть цей, очікуватимуть завантаження й розбору `colors`, щоб мати змогу використовувати залежність.

> Якщо включити цей модуль в файл [`main.js`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/main.js):

Reviewer: ```suggestion
Включімо цей модуль до файлу [`main.js`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/main.js):
```

```js
import colors from "./modules/getColors.js";
import { Canvas } from "./modules/canvas.js";

const circleBtn = document.querySelector(".circle");

// …
```

Застосуймо `colors` замість рядків, що раніше застосовувалися для виклику функцій фігур:

```js
const square1 = new Module.Square(
  myCanvas.ctx,
  myCanvas.listId,
  50,
  50,
  100,
  colors.blue
);

const circle1 = new Module.Circle(
  myCanvas.ctx,
  myCanvas.listId,
  75,
  200,
  100,
  colors.green
);

const triangle1 = new Module.Triangle(
  myCanvas.ctx,
  myCanvas.listId,
  100,
  75,
  190,
  colors.yellow
);
```

Це корисно, бо код у [`main.js`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/main.js) не виконається, поки не завершиться код у [`getColors.js`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/modules/getColors.js). Проте це не завадить завантаженню інших модулів. Наприклад, модуль [`canvas.js`](https://github.com/mdn/js-examples/blob/master/module-examples/top-level-await/modules/canvas.js) завантажуватиметься далі, поки виконується отримання `colors`.

## Написання "ізоморфних" модулів

> Запровадження модулів заохочує екосистему JavaScript поширювати й повторно використовувати код у модульному стилі. Проте це не обов'язково означає, що певний код JavaScript може працювати в будь-якому середовищі. Припустімо, ви знайшли модуль, що породжує геші SHA паролів користувача. Чи можна його використати у браузерному фронтенді? Чи можна його використати на Node.js сервері? Відповідь: буває по різному.

Reviewer: Я здається зрозумів, у чому проблема з "породжує" - українською воно звучить радше як "щось, що створює, причому робить це самодостатньо". Це може підходити для опису `/dev/random` в лінуксах, який буквально є джерелом псевдовипадкових чисел. Проте це погано пасує до операції гешування, яка за визначенням потребує зовнішнього аргумента
```suggestion
Запровадження модулів заохочує екосистему JavaScript поширювати й повторно використовувати код у модульному стилі. Проте це не обов'язково означає, що певний код JavaScript може працювати в будь-якому середовищі. Припустімо, ви знайшли модуль, що обчислює геші SHA паролів користувача. Чи можна його використати у браузерному фронтенді? Чи можна його використати на Node.js сервері? Відповідь: буває по різному.
```

Модулі мають доступ до глобальних змінних, як показано вище. Якщо модуль звертається до глобальних значень, як то `window`, то він може працювати в браузері, але викине помилку на Node.js сервері, адже `window` там немає. Подібно до цього, якщо кодові потрібен для роботи доступ до `process`, то такий код може використовуватись лише в Node.js.

Для максимізації перевикористовності модуля нерідко радять робити код "ізоморфним" – тобто демонструвати однакову поведінку в усіх середовищах виконання. Загалом цього досягають у три способи:

> - Розділення модулів на "ядро" і "зв'язування". У "ядрі" слід зосередитися на щирій логіці JavaScript, як то обчисленні гешу, жодної DOM, звертань до мережі, файлової системи, і відкрити доступ до корисних функцій. У "зв'язуванні" можна зчитувати й змінювати глобальний контекст. Наприклад, "браузерне зв'язування" може вирішити отримати значення з поля введення, а "зв'язування Node" – отримати його з `process.env`, але значення з обох місць передаються до однієї функції ядра й обробляються однаково. Ядро може бути імпортоване в кожному середовищі й використовуватися однаково, тим часом лише зв'язування, котре зазвичай є легковагими, мусить бути специфічним щодо платформи.

Reviewer: ```suggestion
- Розділення модулів на "ядро" і "зв'язування". У "ядрі" слід зосередитися на щирій логіці JavaScript, як то обчисленні гешу, жодної DOM, звертань до мережі, файлової системи – і відкрити доступ до корисних функцій. У "зв'язуванні" можна зчитувати й змінювати глобальний контекст. Наприклад, "браузерне зв'язування" може вирішити отримати значення з поля введення, а "зв'язування Node" – отримати його з `process.env`, але значення з обох місць передаються до однієї функції ядра й обробляються однаково. Ядро може бути імпортоване в кожному середовищі й використовуватися однаково, тим часом лише зв'язування, котре зазвичай є легковагим, мусить бути специфічним щодо платформи.
```
